eth/filters/filter.go

```diff
@@ -261,32 +262,57 @@ func (f *Filter) blockLogs(ctx context.Context, header *types.Header) (logs []*t
 // checkMatches checks if the receipts belonging to the given header contain any log events that
 // match the filter criteria. This function is called when the bloom filter signals a potential match.
 func (f *Filter) checkMatches(ctx context.Context, header *types.Header) (logs []*types.Log, err error) {
+	var (
+		block  *types.Block
+		number = header.Number.Uint64()
+		hash   = header.Hash()
+	)
+	// Fetch block only once and only when its needed
+	lazyLoadBlock := func() (*types.Block, error) {
+		if block != nil {
+			return block, nil
+		}
+		var err error
+		block, err = f.backend.BlockByHash(ctx, hash)
```

Comment:
If possible, please fetch the block using `hash` and `number`, it is more efficient.